File: ios/pocSdkUnico/UnicoSdkModule.swift
import Foundation
import AcessoBio
import React
import AVFoundation

@objc(UnicoSdk)
class UnicoSdkModule: RCTEventEmitter {
    
    private var hasListeners = false
    
    override static func requiresMainQueueSetup() -> Bool {
        return true
    }
    
    override func supportedEvents() -> [String]! {
        return ["onErrorAcessoBio", 
                "onUserClosedCameraManually", 
                "onSystemClosedCameraTimeoutSession", 
                "onSystemChangedTypeCameraTimeoutFaceInference"]
    }
    
    override func startObserving() {
        hasListeners = true
    }
    
    override func stopObserving() {
        hasListeners = false
    }
    
    @objc
    func testConnection(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        DispatchQueue.main.async {
            resolve("Bridge iOS funcionando! ✅")
        }
    }
    
    @objc
    func checkPermissions(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        DispatchQueue.main.async {
            let cameraStatus = AVCaptureDevice.authorizationStatus(for: .video)
            let hasCamera = cameraStatus == .authorized
            
            let result: [String: Any] = [
                "camera": hasCamera,
                "shouldRequest": cameraStatus == .notDetermined
            ]
            
            resolve(result)
        }
    }
    
    @objc
    func captureSelfie(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        DispatchQueue.main.async {
            guard let rootViewController = RCTPresentedViewController() else {
                reject("ERROR", "Não foi possível obter o view controller", nil)
                return
            }
            
            let cameraStatus = AVCaptureDevice.authorizationStatus(for: .video)
            guard cameraStatus == .authorized else {
                reject("PERMISSION_DENIED", "Permissão de câmera necessária. Vá em Configurações > Privacidade > Câmera", nil)
                return
            }
            
            // ✅ CORREÇÃO: Usar a classe que implementa AcessoBioManagerDelegate
            let captureHandler = SelfieCaptureHandler(resolve: resolve, reject: reject, eventEmitter: self)
            captureHandler.startCapture(from: rootViewController)
        }
    }
    
    @objc
    func captureDocument(_ documentType: String, resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        DispatchQueue.main.async {
            guard let rootViewController = RCTPresentedViewController() else {
                reject("ERROR", "Não foi possível obter o view controller", nil)
                return
            }
            
            let cameraStatus = AVCaptureDevice.authorizationStatus(for: .video)
            guard cameraStatus == .authorized else {
                reject("PERMISSION_DENIED", "Permissão de câmera necessária", nil)
                return
            }
            
            // ✅ CORREÇÃO: Usar a classe que implementa AcessoBioManagerDelegate
            let captureHandler = DocumentCaptureHandler(
                documentType: documentType,
                resolve: resolve,
                reject: reject,
                eventEmitter: self
            )
            captureHandler.startCapture(from: rootViewController)
        }
    }
    
    // ✅ REMOVIDOS: Os métodos addListener/removeListeners problemáticos
    // RCTEventEmitter já implementa esses métodos por padrão
}


// ✅ NOVA CLASSE: Handler para captura de selfie baseado no exemplo nativo
class SelfieCaptureHandler: NSObject {
    private let resolve: RCTPromiseResolveBlock
    private let reject: RCTPromiseRejectBlock
    private weak var eventEmitter: RCTEventEmitter?
    private var manager: AcessoBioManager?
    
    init(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock, eventEmitter: RCTEventEmitter?) {
        self.resolve = resolve
        self.reject = reject
        self.eventEmitter = eventEmitter
        super.init()
    }
    
    func startCapture(from viewController: UIViewController) {
        // ✅ CORREÇÃO: Seguindo o exemplo nativo - sem delegate no construtor
        manager = AcessoBioManager(viewController: viewController)
        manager?.setTheme(UnicoTheme())
        manager?.setEnvironment(.UAT)
        
        // Configurar câmera inteligente
        manager?.setSmartFrame(true)
        manager?.setAutoCapture(true)
        
        // Preparar câmera seguindo o exemplo nativo
        manager?.build().prepareSelfieCamera(self, config: UnicoConfig())
    }
}

// MARK: - SelfieCaptureHandler Extensions
extension SelfieCaptureHandler: AcessoBioManagerDelegate {
    func onErrorAcessoBioManager(_ error: ErrorBio!) {
        if let eventEmitter = eventEmitter {
            eventEmitter.sendEvent(withName: "onErrorAcessoBio", body: [
                "code": error.code ?? "UNKNOWN",
                "description": error.description ?? "Erro desconhecido"
            ])
        }
    }
    
    func onUserClosedCameraManually() {
        eventEmitter?.sendEvent(withName: "onUserClosedCameraManually", body: nil)
        reject("USER_CLOSED", "Usuário fechou a câmera", nil)
    }
    
    func onSystemClosedCameraTimeoutSession() {
        eventEmitter?.sendEvent(withName: "onSystemClosedCameraTimeoutSession", body: nil)
        reject("TIMEOUT", "Timeout da sessão", nil)
    }
    
    func onSystemChangedTypeCameraTimeoutFaceInference() {
        eventEmitter?.sendEvent(withName: "onSystemChangedTypeCameraTimeoutFaceInference", body: nil)
    }
}

extension SelfieCaptureHandler: SelfieCameraDelegate {
    func onCameraReady(_ cameraOpener: AcessoBioCameraOpenerDelegate!) {
        cameraOpener.open(self)
    }
    
    func onCameraFailed(_ message: ErrorPrepare!) {
        reject("CAMERA_FAILED", message.description, nil)
    }
}

extension SelfieCaptureHandler: AcessoBioSelfieDelegate {
    func onSuccessSelfie(_ result: SelfieResult!) {
        resolve([
            "base64": result.base64 ?? "",
            "encrypted": result.encrypted ?? ""
        ])
    }
    
    func onErrorSelfie(_ errorBio: ErrorBio!) {
        reject(errorBio.code ?? "ERROR", errorBio.description, nil)
    }
}

// ✅ NOVA CLASSE: Handler para captura de documento baseado no exemplo nativo
class DocumentCaptureHandler: NSObject {
    private let documentType: String
    private let resolve: RCTPromiseResolveBlock
    private let reject: RCTPromiseRejectBlock
    private weak var eventEmitter: RCTEventEmitter?
    private var manager: AcessoBioManager?
    
    init(documentType: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock, eventEmitter: RCTEventEmitter?) {
        self.documentType = documentType
        self.resolve = resolve
        self.reject = reject
        self.eventEmitter = eventEmitter
        super.init()
    }
    
    func startCapture(from viewController: UIViewController) {
        // ✅ CORREÇÃO: Seguindo o exemplo nativo - sem delegate no construtor
        manager = AcessoBioManager(viewController: viewController)
        manager?.setTheme(UnicoTheme())
        manager?.setEnvironment(.UAT)
        
        // Preparar câmera seguindo o exemplo nativo
        manager?.build().prepareDocumentCamera(self, config: UnicoConfig())
    }
    
    private func mapDocumentType(_ documentType: String) -> DocumentEnums {
        switch documentType {
        case "CPF":
            return .CPF
        case "CNH":
            return .CNH
        case "CNH_FRENTE":
            return .cnhFrente
        case "CNH_VERSO":
            return .cnhVerso
        case "RG_FRENTE":
            return .rgFrente
        case "RG_VERSO":
            return .rgVerso
        default:
            return .none
        }
    }
}

// MARK: - DocumentCaptureHandler Extensions
extension DocumentCaptureHandler: AcessoBioManagerDelegate {
    func onErrorAcessoBioManager(_ error: ErrorBio!) {
        if let eventEmitter = eventEmitter {
            eventEmitter.sendEvent(withName: "onErrorAcessoBio", body: [
                "code": error.code ?? "UNKNOWN",
                "description": error.description ?? "Erro desconhecido"
            ])
        }
    }
    
    func onUserClosedCameraManually() {
        eventEmitter?.sendEvent(withName: "onUserClosedCameraManually", body: nil)
        reject("USER_CLOSED", "Usuário fechou a câmera", nil)
    }
    
    func onSystemClosedCameraTimeoutSession() {
        eventEmitter?.sendEvent(withName: "onSystemClosedCameraTimeoutSession", body: nil)
        reject("TIMEOUT", "Timeout da sessão", nil)
    }
    
    func onSystemChangedTypeCameraTimeoutFaceInference() {
        eventEmitter?.sendEvent(withName: "onSystemChangedTypeCameraTimeoutFaceInference", body: nil)
    }
}

extension DocumentCaptureHandler: DocumentCameraDelegate {
    func onCameraReadyDocument(_ cameraOpener: AcessoBioCameraOpenerDelegate!) {
        cameraOpener.openDocument(mapDocumentType(documentType), delegate: self)
    }
    
    func onCameraFailedDocument(_ message: ErrorPrepare!) {
        reject("CAMERA_FAILED", message.description, nil)
    }
}

extension DocumentCaptureHandler: AcessoBioDocumentDelegate {
    func onSuccessDocument(_ result: DocumentResult!) {
        resolve([
            "base64": result.base64 ?? "",
            "encrypted": result.encrypted ?? ""
        ])
    }
    
    func onErrorDocument(_ errorBio: ErrorBio!) {
        reject(errorBio.code ?? "ERROR", errorBio.description, nil)
    }
}

File: android/app/src/main/java/com/pocsdkunico/unico/UnicoSdkModule.kt
package com.pocsdkunico.unico

import android.Manifest
import android.app.Activity
import android.content.pm.PackageManager
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule

import com.acesso.acessobio_android.AcessoBioListener
import com.acesso.acessobio_android.iAcessoBioDocument
import com.acesso.acessobio_android.iAcessoBioSelfie
import com.acesso.acessobio_android.onboarding.AcessoBio
import com.acesso.acessobio_android.onboarding.camera.CameraListener
import com.acesso.acessobio_android.onboarding.camera.UnicoCheckCameraOpener
import com.acesso.acessobio_android.onboarding.camera.document.DocumentCameraListener
import com.acesso.acessobio_android.onboarding.models.Environment
import com.acesso.acessobio_android.onboarding.types.DocumentType
import com.acesso.acessobio_android.services.dto.ErrorBio
import com.acesso.acessobio_android.services.dto.ResultCamera

class UnicoSdkModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    
    companion object {
        const val NAME = "UnicoSdk"
        private const val TAG = "UnicoSdkModule"
        private const val CAMERA_PERMISSION_REQUEST_CODE = 1001
    }

    private var currentPromise: Promise? = null
    private val activity: Activity? get() = currentActivity

    override fun getName(): String = NAME

    private fun runOnMainThread(action: () -> Unit) {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            action()
        } else {
            Handler(Looper.getMainLooper()).post(action)
        }
    }


    private fun hasPermissions(): Boolean {
        val activity = activity ?: return false
        return ContextCompat.checkSelfPermission(
            activity, 
            Manifest.permission.CAMERA
        ) == PackageManager.PERMISSION_GRANTED
    }

    @ReactMethod
    fun checkPermissions(promise: Promise) {
        try {
            val hasCamera = hasPermissions()
            val resultMap = Arguments.createMap().apply {
                putBoolean("camera", hasCamera)
                putBoolean("shouldRequest", !hasCamera)
            }
            promise.resolve(resultMap)
        } catch (e: Exception) {
            promise.reject("ERROR", e.message)
        }
    }

    @ReactMethod
    fun testConnection(promise: Promise) {
        runOnMainThread {
            try {
                promise.resolve("Bridge Android funcionando! Activity: ${activity != null}")
            } catch (e: Exception) {
                promise.reject("ERROR", e.message ?: "Test failed")
            }
        }
    }

    @ReactMethod
    fun captureSelfie(promise: Promise) {
        Log.d(TAG, "=== CAPTURA SELFIE INICIADA ===")
        currentPromise = promise
        
        runOnMainThread {
            try {
                val activity = activity ?: run {
                    Log.e(TAG, "Activity é null")
                    currentPromise?.reject("ERROR", "Activity not found")
                    currentPromise = null
                    return@runOnMainThread
                }

          
                if (!hasPermissions()) {
                    Log.e(TAG, "❌ Permissão de câmera não concedida")
                    currentPromise?.reject(
                        "PERMISSION_DENIED", 
                        "Permissão de câmera necessária. Vá em Configurações > Apps > Permissões > Câmera e ative."
                    )
                    currentPromise = null
                    return@runOnMainThread
                }

                Log.d(TAG, "✅ Permissão de câmera OK")

                val config = UnicoConfig()
                Log.d(TAG, "Bundle Identifier: '${config.getBundleIdentifier()}'")
                Log.d(TAG, "Package name real: ${activity.packageName}")

                val callback = object : AcessoBioListener {
                    override fun onErrorAcessoBio(errorBio: ErrorBio?) {
                        Log.e(TAG, "❌ onErrorAcessoBio - Code: ${errorBio?.code}, Description: ${errorBio?.description}")
                        sendEvent("onErrorAcessoBio", createErrorMap(errorBio))
                        currentPromise?.reject("SDK_ERROR", "Erro SDK: ${errorBio?.description}")
                        currentPromise = null
                    }

                    override fun onUserClosedCameraManually() {
                        Log.d(TAG, "👤 onUserClosedCameraManually")
                        sendEvent("onUserClosedCameraManually", null)
                        currentPromise?.reject("USER_CANCELLED", "User cancelled the capture")
                        currentPromise = null
                    }

                    override fun onSystemClosedCameraTimeoutSession() {
                        Log.d(TAG, "⏰ onSystemClosedCameraTimeoutSession")
                        sendEvent("onSystemClosedCameraTimeoutSession", null)
                        currentPromise?.reject("TIMEOUT", "Timeout exceeded")
                        currentPromise = null
                    }

                    override fun onSystemChangedTypeCameraTimeoutFaceInference() {
                        Log.d(TAG, "🔄 onSystemChangedTypeCameraTimeoutFaceInference")
                        sendEvent("onSystemChangedTypeCameraTimeoutFaceInference", null)
                    }
                }

                val cameraListener = object : iAcessoBioSelfie {
                    override fun onSuccessSelfie(result: ResultCamera?) {
                        Log.d(TAG, "✅ onSuccessSelfie - resultado recebido")
                        result?.let {
                            val resultMap = Arguments.createMap().apply {
                                putString("base64", it.base64)
                                putString("encrypted", it.encrypted)
                            }
                            currentPromise?.resolve(resultMap)
                        } ?: run {
                            Log.e(TAG, "❌ Resultado da selfie é null")
                            currentPromise?.reject("ERROR", "Null result")
                        }
                        currentPromise = null
                    }

                    override fun onErrorSelfie(errorBio: ErrorBio?) {
                        Log.e(TAG, "❌ onErrorSelfie - Code: ${errorBio?.code}, Description: ${errorBio?.description}")
                        currentPromise?.reject("SELFIE_ERROR", errorBio?.description ?: "Error capturing selfie")
                        currentPromise = null
                    }
                }

                val selfieCallback = object : CameraListener {
                    override fun onCameraReady(cameraOpener: UnicoCheckCameraOpener.Camera) {
                        Log.d(TAG, "📸 Camera pronta, abrindo...")
                        cameraOpener.open(cameraListener)
                    }

                    override fun onCameraFailed(message: String?) {
                        Log.e(TAG, "❌ Camera falhou: $message")
                        currentPromise?.reject("CAMERA_FAILED", message ?: "Failed to prepare camera")
                        currentPromise = null
                    }
                }

                Log.d(TAG, "🔧 Construindo AcessoBio com Environment.UAT...")
                
        
                val theme = UnicoTheme()
                
                AcessoBio(activity, callback)
                    .setAutoCapture(true)
                    .setSmartFrame(true)
                    .setEnvironment(Environment.UAT)
                    .setTheme(theme) 
                    .build()
                    .prepareCamera(config, selfieCallback)
                    
                Log.d(TAG, "✅ AcessoBio.prepareCamera() chamado com sucesso")

            } catch (e: Exception) {
                Log.e(TAG, "❌ Exception geral em captureSelfie", e)
                currentPromise?.reject("ERROR", "Erro geral: ${e.message}")
                currentPromise = null
            }
        }
    }

    @ReactMethod
    fun captureDocument(documentType: String, promise: Promise) {
        Log.d(TAG, "captureDocument called with type: $documentType")
        currentPromise = promise
        
        runOnMainThread {
            try {
                val activity = activity ?: run {
                    currentPromise?.reject("ERROR", "Activity not found")
                    currentPromise = null
                    return@runOnMainThread
                }

                if (!hasPermissions()) {
                    currentPromise?.reject(
                        "PERMISSION_DENIED", 
                        "Permissão de câmera necessária"
                    )
                    currentPromise = null
                    return@runOnMainThread
                }

                val callback = object : AcessoBioListener {
                    override fun onErrorAcessoBio(errorBio: ErrorBio?) {
                        sendEvent("onErrorAcessoBio", createErrorMap(errorBio))
                        currentPromise?.reject("SDK_ERROR", errorBio?.description ?: "SDK Error")
                        currentPromise = null
                    }

                    override fun onUserClosedCameraManually() {
                        sendEvent("onUserClosedCameraManually", null)
                        currentPromise?.reject("USER_CANCELLED", "User cancelled the capture")
                        currentPromise = null
                    }

                    override fun onSystemClosedCameraTimeoutSession() {
                        sendEvent("onSystemClosedCameraTimeoutSession", null)
                        currentPromise?.reject("TIMEOUT", "Timeout exceeded")
                        currentPromise = null
                    }

                    override fun onSystemChangedTypeCameraTimeoutFaceInference() {
                        sendEvent("onSystemChangedTypeCameraTimeoutFaceInference", null)
                    }
                }

                val cameraListener = object : iAcessoBioDocument {
                    override fun onSuccessDocument(result: ResultCamera?) {
                        result?.let {
                            val resultMap = Arguments.createMap().apply {
                                putString("base64", it.base64)
                                putString("encrypted", it.encrypted)
                            }
                            currentPromise?.resolve(resultMap)
                        } ?: currentPromise?.reject("ERROR", "Null result")
                        currentPromise = null
                    }

                    override fun onErrorDocument(errorMessage: String?) {
                        currentPromise?.reject("ERROR", errorMessage ?: "Error capturing document")
                        currentPromise = null
                    }
                }

                val docType = when (documentType) {
                    "CPF" -> DocumentType.CPF
                    "CNH" -> DocumentType.CNH
                    "CNH_FRENTE" -> DocumentType.CNH_FRENTE
                    "CNH_VERSO" -> DocumentType.CNH_VERSO
                    "RG_FRENTE" -> DocumentType.RG_FRENTE
                    "RG_VERSO" -> DocumentType.RG_VERSO
                    else -> DocumentType.NONE
                }

                val documentCallback = object : DocumentCameraListener {
                    override fun onCameraReady(cameraOpener: UnicoCheckCameraOpener.Document?) {
                        cameraOpener?.open(docType, cameraListener)
                    }

                    override fun onCameraFailed(message: String?) {
                        currentPromise?.reject("CAMERA_FAILED", message ?: "Failed to prepare camera")
                        currentPromise = null
                    }
                }

                val theme = UnicoTheme()

                AcessoBio(activity, callback)
                    .setAutoCapture(true)
                    .setSmartFrame(true)
                    .setEnvironment(Environment.UAT)
                    .setTheme(theme)
                    .build()
                    .prepareDocumentCamera(UnicoConfig(), documentCallback)

            } catch (e: Exception) {
                Log.e(TAG, "Exception in captureDocument", e)
                currentPromise?.reject("ERROR", e.message ?: "Unknown error")
                currentPromise = null
            }
        }
    }

    private fun createErrorMap(errorBio: ErrorBio?): WritableMap {
        return Arguments.createMap().apply {
            putString("code", errorBio?.code?.toString() ?: "UNKNOWN")
            putString("description", errorBio?.description ?: "Unknown error")
        }
    }

    private fun sendEvent(eventName: String, params: WritableMap?) {
        try {
            reactApplicationContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                .emit(eventName, params)
        } catch (e: Exception) {
            Log.e(TAG, "Error sending event: $eventName", e)
        }
    }

    @ReactMethod
    fun addListener(eventName: String) {
        // Necessário para RCTEventEmitter
    }

    @ReactMethod
    fun removeListeners(count: Int) {
        // Necessário para RCTEventEmitter
    }
}